### DISCUSSION TITLE
A 'native' parser described via some parser grammatic of choice (Yacc, Antlr, maybe some parser-combinator?)

### DISCUSSION THREAD

**Discussion by utrack** [a-h/templ#936] [Ideas]
Basically the title :) What do you think about creating a native parser that won't neccessarily use the current astparse hacks?
I think it would be more robust and the changes can be made easier.

If you're interested - I can take this to development and eventually, send MRs.

**Comment by a-h** [comment]
My main concerns around this type of thing would be:

* Ability to produce good error messages - my previous experiments with use parser generators have resulted in poor error messages, and lack of token positions and useful diagnostics to feed into the LSPs diagnostics
* Ability to support LSP behaviour - need more than "unexpected token", need to understand relationships etc.
* Complexity of the solution and long term support of it - the current design is approachable, people can work with it and understand it to make changes
* Ability to test the solution to prevent accidental regressions

A formal grammar that describes templ could be useful, if it could be used to generate the sytnax highlighting grammars for the various editors, since that is a regular pain point.

At the moment, I don't see it as a main priority though.
